From 6d94c3ee14eb0d3c4268e9ac770f42386ffeb037 Mon Sep 17 00:00:00 2001
From: Anna REITER <annareiter001@gmail.com>
Date: Wed, 26 Aug 2026 10:32:40 +0200
Subject: [PATCH 1/3] rvvts: DBT-RISE-RISCV: fix vcsr write

The RISC-V Vector specification defines vcsr as an alias for vxrm and vxsat,
so writes to vcsr must update both fields consistently.

This bug was found by the open-source RISC-V Vector Test Suite (RVVTS),
developed at the Institute for Complex Systems, JKU Linz:
https://github.com/ics-jku/RVVTS

Signed-off-by: Anna REITER <annareiter001@gmail.com>
Signed-off-by: Manfred SCHLAEGL <manfred.schlaegl@gmx.at>
---
 src/vm/interp/vm_rv32gcv.cpp | 5 +++++
 src/vm/interp/vm_rv64gcv.cpp | 5 +++++
 2 files changed, 10 insertions(+)

diff --git a/src/vm/interp/vm_rv32gcv.cpp b/src/vm/interp/vm_rv32gcv.cpp
index 087bf07..e78bede 100644
--- a/src/vm/interp/vm_rv32gcv.cpp
+++ b/src/vm/interp/vm_rv32gcv.cpp
@@ -2772,6 +2772,11 @@ typename vm_base<ARCH>::virt_addr_t vm_impl<ARCH>::execute_inst(finish_cond_e co
                                             *(X+rd) = xrd;
                                         }
                                     }
+
+                                    if(csr == 0xF) { // vcsr
+                                        super::template write_mem<uint32_t>(traits::CSR, 0xa, (uint64_t) zimm >> 1); // vxrm
+                                        super::template write_mem<uint32_t>(traits::CSR, 0x9, (uint64_t) zimm & 1); // vxsat
+                                    }
                                 }
                     break;
                 }// @suppress("No break at end of case")
diff --git a/src/vm/interp/vm_rv64gcv.cpp b/src/vm/interp/vm_rv64gcv.cpp
index 7922f49..d21c416 100644
--- a/src/vm/interp/vm_rv64gcv.cpp
+++ b/src/vm/interp/vm_rv64gcv.cpp
@@ -3166,6 +3166,11 @@ typename vm_base<ARCH>::virt_addr_t vm_impl<ARCH>::execute_inst(finish_cond_e co
                                             *(X+rd) = xrd;
                                         }
                                     }
+
+                                    if(csr == 0xF) { // vcsr
+                                        super::template write_mem<uint64_t>(traits::CSR, 0xa, (uint64_t) zimm >> 1); // vxrm
+                                        super::template write_mem<uint64_t>(traits::CSR, 0x9, (uint64_t) zimm & 1); // vxsat
+                                    }
                                 }
                     break;
                 }// @suppress("No break at end of case")
